### PR TITLE
autonomi_cli: add evm wallet create, balance

### PR DESCRIPTION
**Pull request by ermineJose**
### Description

Create a EVM wallet through autonomi CLI.

### Related Issue

Fixes #<issue_number> (if applicable).

### Type of Change

Please mark the types of changes made in this pull request.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please describe):

### Checklist

Please ensure all of the following tasks have been completed:

- [ ] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have followed the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) guidelines for commit messages.
- [ ] I have verified my commit messages with [commitlint](https://commitlint.js.org/#/).

### CODE REVIEW

File: autonomi-cli/src/commands/wallet.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+use autonomi::Multiaddr;
+use autonomi::wallet::*;
+use color_eyre::{
+    eyre::{bail, Context},
+    Result,
+};
+
+
+pub fn create(peers: Vec<Multiaddr>) -> Result<()> {
```
Comment (RolandSherwin) [inline autonomi-cli/src/commands/wallet.rs]: I don't think we'd need `peers` argument for `create/balance`. 
`balance` uses just the evm network. the `peers` is used to bootstrap into OUR network (which we don't have to, for these commands). `create` does not need any network.

File: sn_evm/src/wallet.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+use std::io::Write;
+use std::path::Path;
+
+pub fn get_random_private_key() -> String {
+    get_random_private_key_for_wallet()
+}
+
+pub fn create_a_evm_wallet(private_key: &String) -> String {
+    let network = get_evm_network_from_env()
+                        .expect("Failed to get EVM network from environment variables");
```
Comment (RolandSherwin) [inline sn_evm/src/wallet.rs]: `expect` will cause the program to panic/crash. We should return an error here. 
Same at all the places where we use `expect`

File: autonomi-cli/src/commands/wallet.rs (hunk 1 above)
Comment (grumbach) [reply autonomi-cli/src/commands/wallet.rs]: Agreed

File: sn_evm/src/wallet.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+                                                .expect("Could not init deployer wallet");
+    hex::encode(wallet.address())
+}
+
+pub fn create_file_with_keys(private_key: String, public_key: String) -> String {
+    let mut file = File::create(public_key.clone()).expect("could not create file");
+    file.write_all(private_key.as_bytes()).expect("Not able to write into file");
+    let file_path = Path::new(&public_key).canonicalize().expect("Not able to find the absolute path for the file");
+    file_path.to_string_lossy().to_string()
+}
```
Comment (grumbach) [inline sn_evm/src/wallet.rs]: Same here, return errors instead of expect

File: sn_evm/src/wallet.rs (hunk 2 above)
Comment (grumbach) [reply sn_evm/src/wallet.rs]: Agreed too

File: autonomi-cli/src/access/data_dir.rs [hunk 4]
```diff
@@ -16,4 +16,4 @@ pub fn get_client_data_dir_path() -> Result<PathBuf> {
     home_dirs.push("autonomi");
     std::fs::create_dir_all(home_dirs.as_path()).wrap_err("Failed to create data dir")?;
     Ok(home_dirs)
-}
+}
```
Comment (grumbach) [inline autonomi-cli/src/access/data_dir.rs:19]: Maybe that change is not necessary?

File: autonomi-cli/src/commands.rs [hunk 5]
```diff
@@ -163,5 +188,16 @@ pub async fn handle_subcommand(opt: Opt) -> Result<()> {
...
             VaultCmd::Sync => vault::sync(peers.await?),
         },
+        SubCmd::Wallet { command } => match command {
+            WalletCmd::Create {
+                encrypt,
+                password,
+                private_key,
+
+            } => {
+                wallet::create( encrypt, password,private_key)
```
Comment (grumbach) [inline autonomi-cli/src/commands.rs:198]: Does this spacing pass clippy?

File: sn_evm/Cargo.toml [hunk 6]
```diff
@@ -31,6 +31,10 @@ tracing = { version = "~0.1.26" }
 xor_name = "5.0.0"
 ring = "0.17.8"
 tempfile = "3.10.1"
+tokio = { version = "1", features = ["full"] }
+rpassword = "7.0"
+color-eyre = "~0.6"
+dirs-next = "~2.0.0"
```
Comment (grumbach) [inline sn_evm/Cargo.toml:37]: Why do we need dirs next here? 
Better keep disk stuff out of this lib for wasm compatibility

File: sn_evm/src/error.rs [hunk 7]
```diff
@@ -30,4 +30,13 @@ pub enum EvmError {
...
     InvalidQuotePublicKey,
+
+    #[error("Failed to encrypt secret key: {0}")]
+    FailedToEncryptKey(String),
+
+    #[error("Failed to decrypt secret key: {0}")]
+    FailedToDecryptKey(String),
+
+    #[error("Failed to process Operation: {0}")]
+    OperationError(String),
```
Comment (grumbach) [inline sn_evm/src/error.rs:41]: These errors should be a separate type

File: sn_evm/src/lib.rs [hunk 8]
```diff
@@ -25,6 +25,8 @@ pub use evmlib::Network as EvmNetwork;
 mod amount;
 mod data_payments;
 mod error;
+pub mod wallet;
+pub mod encryption;
```
Comment (grumbach) [inline sn_evm/src/lib.rs:29]: The modules should not be pub. Export the needed types only.

File: autonomi/src/wallet.rs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                        },
+                        None => {
+                            println!("The provided Password is not proper, can not proceed further");
+                        }
+                    }
+                    
+                }
+            }
+        },
+        None => {
```
Comment (grumbach) [inline autonomi/src/wallet.rs:87]: Let's try to keep reasonable indentation level. I try to keep it under 3 levels.

**Review by grumbach — CHANGES_REQUESTED** [review]
Left some comments, not sure we want wallet integration for release so blocking this for now.

**Comment by mickvandijke** [comment]
Closed by https://github.com/maidsafe/safe_network/pull/2352

**Pull request closed without merging**